domain/agentbinary/service/store.go

```diff
@@ -119,13 +144,41 @@ func (s *AgentBinaryStore) Add(
 // It always overwrites the binary in the store and the metadata in the database for the
 // given version and arch if it already exists.
 // It accepts the SHA256 hash of the binary.
-// It returns [coreerrors.NotSupported] if the architecture is not found in the database.
-// It returns [coreerrors.NotFound] if object store UUID is not found in the database.
-// It returns [coreerrors.NotValid] if the agent version is not valid.
+// The following errors can be returned:
+// - [coreerrors.NotSupported] if the architecture is not supported.
+// - [github.com/juju/juju/domain/agentbinary/errors.AlreadyExists] if an agent
+// binary already exists for this version and architecture.
+// - [github.com/juju/juju/domain/agentbinary/errors.ObjectNotFound] if there
+// was a problem referencing the agent binary metadata with the previously saved
+// binary object. This error should be considered an internal problem. It is
+// discussed here to make the caller aware of future problems.
+// - [coreerrors.NotValid] if the agent version is not valid or the SHA256 hash doesn't match the generated hash.
 func (s *AgentBinaryStore) AddWithSHA256(
-	ctx context.Context, r io.Reader, version coreagentbinary.Version, size int64, sha256 string,
+	ctx context.Context, r io.Reader,
+	version coreagentbinary.Version,
+	size int64, sha256 string,
 ) (resultErr error) {
-	// TODO: validates the sha256 and generate a new SHA384 in the store. It will be implemented in JUJU-7734.
-	sha384 := sha256
-	return s.Add(ctx, r, version, size, sha384)
+	if err := version.Validate(); err != nil {
+		return errors.Errorf("agent version %q is not valid: %w", version, err)
+	}
+
+	// Ideally, we should pass the SHA256 hash to the object store
+	// and let it verify the hash. But the object store doesn't support
+	// this yet. So we have to calculate the hash ourselves.
+	hasher256 := cryptosha256.New()
+	hasher384 := cryptosha512.New384()
+	if _, err := io.Copy(io.MultiWriter(hasher256, hasher384), r); err != nil {
+		return errors.Errorf("generating SHA for agent binary %q: %w", version, err)
+	}
+
+	encoded256 := hex.EncodeToString(hasher256.Sum(nil))
+	encoded384 := hex.EncodeToString(hasher384.Sum(nil))
+	if sha256 != encoded256 {
+		return errors.Errorf(
+			"SHA256 mismatch for agent binary %q: expected %q, got %q",
+			version, sha256, encoded256,
+		).Add(coreerrors.NotValid)
+	}
+
+	return s.add(ctx, r, version, size, encoded384)
```

Comment:
We've just eaten everything from the reader. We need to buffer to a temporary file, then pass that file to `add` then delete it when we return.

Reply:
Damn, Fixed!
Thanks!